app/helpers/fhir_helper.rb

```diff
@@ -232,6 +232,65 @@ def close_contact_from_fhir(related_person)
     }
   end
 
+  # Returns a representative FHIR::Immunization for an instance of a Sara Alert Vaccine.
+  # https://www.hl7.org/fhir/immunization.html
+  def vaccine_as_fhir(vaccine)
+    FHIR::Immunization.new(
+      meta: FHIR::Meta.new(lastUpdated: vaccine.updated_at.strftime('%FT%T%:z')),
+      id: vaccine.id,
+      status: 'completed',
+      vaccineCode: [
+        FHIR::CodeableConcept.new(
+          text: vaccine.product_name,
+          coding: Vaccine.product_codes_by_name(vaccine.group_name, vaccine.product_name).map do |code|
+            FHIR::Coding.new(code: code['code'], system: code['system'])
+          end
+        )
+      ],
+      patient: FHIR::Reference.new(reference: "Patient/#{vaccine.patient_id}"),
+      occurrenceDateTime: vaccine.administration_date,
+      note: FHIR::Annotation.new(text: vaccine.notes),
+      protocolApplied: [
+        {
+          targetDisease: [
+            FHIR::CodeableConcept.new(
+              text: vaccine.group_name,
+              coding: Vaccine.group_codes_by_name(vaccine.group_name).map do |code|
+                FHIR::Coding.new(code: code['code'], system: code['system'])
+              end
+            )
+          ],
+          doseNumberString: vaccine.dose_number
+        }
+      ]
+    )
+  end
+
+  # Create a hash of atttributes that corresponds to a Sara Alert Patient (and can be used to
+  # create new ones, or update existing ones), using the given FHIR::Patient.
+  # Hash is of the form:
+  # {
+  #  attribute_name: { value: <converted-value>, path: <fhirpath-to-corresponding-fhir-element>, errors: <array-of-messages> }
+  # }
+  def vaccine_from_fhir(vaccine)
+    group_coding = vaccine&.protocolApplied&.first&.targetDisease&.first&.coding&.first
+    group = Vaccine.group_name_by_code(group_coding&.system, group_coding&.code)
+    group_errors = ['not in required ValueSet'] if group.nil? && !group_coding.nil?
+
+    product_name_coding = vaccine&.vaccineCode&.first&.coding&.first
+    product_name = Vaccine.product_name_by_code(group, product_name_coding&.system, product_name_coding&.code)
+    product_name_errors = ['not in required ValueSet'] if product_name.nil? && !product_name_coding.nil?
```

Comment:
>I don't love the approach I took here, but I think this kind of error messaging is important. Without it, the user will see a message that says their code (say 206) is wrong, and that Sara Alert expects something like "Moderna COVID-19 Vaccine". But, the user should not send in "Moderna COVID-19 Vaccine" using FHIR, they should send the code that gets mapped to that concept.

What about instead adding a validation on `Vaccine` scoped to `:api`, and using a different error message to include the list of valid codes? That way you could also add a `presence` validation, because as of now if you exclude a necessary code, you get the same message "not in required ValueSet".